src/os/bluestore/BlueStore.cc

```diff
@@ -5599,21 +5599,30 @@ void BlueStore::_do_write_big(
     bufferlist::iterator& blp,
     WriteContext *wctx)
 {
+  uint64_t max_blob_len = length;
+  if (wctx->compress) {
+    max_blob_len = MIN(length, wctx->comp_blob_size);
+  }
   dout(10) << __func__ << " 0x" << std::hex << offset << "~0x" << length
+	   << " max_blob_len 0x" << max_blob_len
 	   << std::dec << dendl;
-  int64_t blob;
-  bluestore_blob_t *b = o->onode.add_blob(&blob);
-  b->length = length;
-  wctx->blob_new.push_back(b);
-  bufferlist t;
-  blp.copy(length, t);
-  wctx->bl_new.push_back(t);
-  o->onode.punch_hole(offset, length, &wctx->lex_old);
-  o->onode.extent_map[offset] = bluestore_lextent_t(blob, 0, length, 0);
-  b->ref_map.get(0, length);
-  dout(20) << __func__ << "  lex 0x" << std::hex << offset << std::dec << ": "
-	   << o->onode.extent_map[offset] << dendl;
-  dout(20) << __func__ << "  blob " << *b << dendl;
+  while (length > 0) {
+    int64_t blob;
+    bluestore_blob_t *b = o->onode.add_blob(&blob);
+    wctx->blob_new.push_back(b);
+    b->length = MIN(max_blob_len, length);
+    bufferlist t;
+    blp.copy(b->length, t);
+    wctx->bl_new.push_back(t);
```

Comment:
One can probably do compress here and thus eliminate the additional copy and reduce memory footprint a bit.
Another option is to store <start_iterator, end_iterator> pair instead of  bufferlist copy in bl_new.